### PR TITLE
Add exponential backoff for retries after error

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: src/DataAggregator/GlobalWorkers/LoopedWorkerBase.cs [hunk 1]
```diff
@@ -75,6 +75,72 @@ public enum BehaviourOnFault
...
+    private readonly int _exponentialBackoffErrorsGracePeriod;
+    private readonly uint _delayAfterErrorExponentialRate;
+    private readonly uint _delayAfterErrorMaxExponent;
+
+    public ExponentialBackoffDelayBetweenLoopsStrategy(
+        TimeSpan minDelayAfterSuccess,
+        TimeSpan baseDelayAfterError,
+        int exponentialBackoffErrorsGracePeriod,
+        uint delayAfterErrorExponentialRate,
+        uint delayAfterErrorMaxExponent
```
Comment (dhedey) [inline src/DataAggregator/GlobalWorkers/LoopedWorkerBase.cs:107]: I feel like a maxDelay might be easier to sense-check without understanding the inner workings of this class 👍

File: src/DataAggregator/NodeScopedWorkers/NodeMempoolFullTransactionReaderWorker.cs [hunk 2]
```diff
@@ -86,6 +87,12 @@ namespace DataAggregator.NodeScopedWorkers;
 /// </summary>
 public class NodeMempoolFullTransactionReaderWorker : NodeWorker
 {
+    private static readonly IDelayBetweenLoopsStrategy _delayBetweenLoopsStrategy =
+        new ExponentialBackoffDelayBetweenLoopsStrategy(
+            TimeSpan.FromMilliseconds(500),
+            TimeSpan.FromMilliseconds(500),
+            1, 2, 5);
```
Comment (dhedey) [inline src/DataAggregator/NodeScopedWorkers/NodeMempoolFullTransactionReaderWorker.cs:94]: I might be tempted to add in the optional prefixes to variable names to explain what these numbers mean :)

EG:
```
new ExponentialBackoffDelayBetweenLoopsStrategy(
    minDelayAfterSuccess: X,
    ...
)
```

File: src/DataAggregator/GlobalWorkers/LoopedWorkerBase.cs [hunk 3]
```diff
@@ -75,6 +75,72 @@ public enum BehaviourOnFault
...
+public class ExponentialBackoffDelayBetweenLoopsStrategy : IDelayBetweenLoopsStrategy
+{
+    private readonly TimeSpan _minDelayAfterSuccess;
+    private readonly TimeSpan _baseDelayAfterError;
+    private readonly int _exponentialBackoffErrorsGracePeriod;
+    private readonly uint _delayAfterErrorExponentialRate;
+    private readonly uint _delayAfterErrorMaxExponent;
+
+    public ExponentialBackoffDelayBetweenLoopsStrategy(
+        TimeSpan minDelayAfterSuccess,
```
Comment (dhedey) [inline src/DataAggregator/GlobalWorkers/LoopedWorkerBase.cs:103]: This is actually `maxDelayAfterSuccess`, or perhaps `delayBetweenLoopTriggersIfSuccessful`

**Review by dhedey — COMMENTED** [review]
Nice stuff! Some minor tweak suggestions, but looks good

File: src/DataAggregator/GlobalWorkers/MempoolPrunerWorker.cs [hunk 4]
```diff
@@ -71,13 +71,18 @@ namespace DataAggregator.GlobalWorkers;
 /// </summary>
 public class MempoolPrunerWorker : GlobalWorker
 {
+    private static readonly IDelayBetweenLoopsStrategy _delayBetweenLoopsStrategy =
+        IDelayBetweenLoopsStrategy.ConstantDelayStrategy(
+            TimeSpan.FromMilliseconds(30),
+            TimeSpan.FromMilliseconds(10));
```
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply src/DataAggregator/GlobalWorkers/MempoolPrunerWorker.cs:77]: good catch

File: src/DataAggregator/NodeScopedWorkers/NodeMempoolFullTransactionReaderWorker.cs (hunk 2 above)
Comment (LukasGasior1) [reply src/DataAggregator/NodeScopedWorkers/NodeMempoolFullTransactionReaderWorker.cs:94]: Rider IDE adds it automatically (which is a bit annoying). But yeah, named parameters wouldn't hurt here for anyone using a different IDE.

File: src/DataAggregator/Workers/GlobalWorkers/LoopedWorkerBase.cs [hunk 5]
```diff
@@ -67,7 +67,7 @@
 using NodaTime;
 using System.Diagnostics;
 
-namespace DataAggregator.GlobalWorkers;
+namespace DataAggregator.Workers.GlobalWorkers;
```
Comment (dhedey) [inline src/DataAggregator/Workers/GlobalWorkers/LoopedWorkerBase.cs:70]: This file should be in the parent Workers folder now :)

**Review by dhedey — APPROVED** [review]

**Review by dhedey — APPROVED** [review]

**Pull request merged**